src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -365,7 +365,23 @@ trait TypeComparers {
 
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
-    def isSub(ntp1: Type, ntp2: Type) = (ntp1.withoutAnnotations, ntp2.withoutAnnotations) match {
+
+    def isSub(tp1: Type, tp2: Type) = (tp1.withoutAnnotations, tp2.withoutAnnotations) match {
+      case (tv1 @ TypeVar(_, _), tv2 @ TypeVar(_, _)) if settings.YhkTypevarUnification =>
+        (tv2.params corresponds tv1.params)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv1.addHiBound(tv2); tv2.addLoBound(tv1); true }
```

Comment:
Will investigate.

Reply:
I tried to add `assert(false)` to this branch and it is not triggered by tests or bootstrap. What do you suggest? Should I just change that branch to `assert(false)`?

Reply:
I wouldn't risk a crash, but we could consider a warning to test our theory since this codepath isn't enabled by default. I haven't thought about it much, but there's precedent (grep "please report" for another warning).

Reply:
Done.

Reply:
We can get here thanks to https://github.com/scala/scala/blob/3a76f41d531a2aefba5c605f5b1fb738c33c5fa6/src/compiler/scala/tools/nsc/typechecker/Infer.scala#L538-L550